File: sdks/python/packages/sudomimus-token/src/sudomimus_token/id_token.py
"""OIDC ``id_token`` parsing and verification.

An id_token is a standard OIDC JWT: its claims live in the body and it is
signed by the **platform** key (resolve it from the OIDC JWKS by the header
``kid``), not by an application's signing key. This module therefore checks
``exp`` from the body, not the header.
"""

from __future__ import annotations

import binascii
from dataclasses import dataclass
from datetime import UTC, datetime

from pydantic import ValidationError

from ._codec import decode_base64url, verify_rs256
from .errors import TokenError, TokenErrorCode
from .models import IdTokenBody, IdTokenHeader


@dataclass(frozen=True, slots=True)
class IdToken:
    """A parsed OIDC id_token (header + body + verbatim signing input)."""

    raw: str
    signing_input: bytes
    signature: bytes
    header: IdTokenHeader
    body: IdTokenBody

    def verify_signature(self, public_key_pem: str) -> bool:
        """Return ``True`` when the signature matches the given RSA public key."""
        return verify_rs256(public_key_pem, self.signing_input, self.signature)


def parse_id_token(jwt: str) -> IdToken:
    """Parse an id_token into its header and body without verifying it."""
    if not jwt:
        raise TokenError(TokenErrorCode.INVALID_JWT, "Token is empty.")
    parts = jwt.split(".")
    if len(parts) != 3:
        raise TokenError(
            TokenErrorCode.INVALID_JWT,
            f"Token must have exactly three dot-separated segments; got {len(parts)}.",
        )
    header_segment, body_segment, signature_segment = parts
    try:
        header_bytes = decode_base64url(header_segment)
        body_bytes = decode_base64url(body_segment)
        signature_bytes = decode_base64url(signature_segment)
    except (binascii.Error, ValueError) as exc:
        raise TokenError(
            TokenErrorCode.INVALID_JWT, f"Failed to decode JWT segments: {exc}"
        ) from exc

    try:
        header = IdTokenHeader.model_validate_json(header_bytes)
        body = IdTokenBody.model_validate_json(body_bytes)
    except ValidationError as exc:
        raise TokenError(
            TokenErrorCode.INVALID_JWT, f"Failed to deserialize id_token: {exc}"
        ) from exc

    return IdToken(
        raw=jwt,
        signing_input=f"{header_segment}.{body_segment}".encode("ascii"),
        signature=signature_bytes,
        header=header,
        body=body,
    )


def verify_id_token(
    jwt: str,
    platform_public_key_pem: str,
    *,
    audience: str | None = None,
    issuer: str | None = None,
    nonce: str | None = None,
    now: datetime | None = None,
) -> IdToken:
    """Verify an OIDC id_token against a platform public key.

    Checks the body ``exp``, the RS256 signature, and any supplied
    ``audience`` / ``issuer`` / ``nonce`` expectations. Returns the parsed
    token or raises :class:`TokenError`.
    """
    parsed = parse_id_token(jwt)
    current = now if now is not None else datetime.now(tz=UTC)

    if datetime.fromtimestamp(parsed.body.exp, tz=UTC) <= current:
        raise TokenError(TokenErrorCode.EXPIRED, "id_token has expired.")

    if not parsed.verify_signature(platform_public_key_pem):
        raise TokenError(
            TokenErrorCode.INVALID_SIGNATURE,
            "id_token signature does not match the platform public key.",
        )

    if audience is not None and parsed.body.aud != audience:
        raise TokenError(
            TokenErrorCode.WRONG_AUDIENCE,
            "id_token `aud` does not match the expected client id.",
        )

    if issuer is not None and parsed.body.iss != issuer:
        raise TokenError(
            TokenErrorCode.WRONG_ISSUER,
            "id_token `iss` does not match the expected issuer.",
        )

    if nonce is not None and parsed.body.nonce != nonce:
        raise TokenError(
            TokenErrorCode.WRONG_NONCE,
            "id_token `nonce` does not match the value sent at /authorize.",
        )

    return parsed

File: sdks/java/token/src/main/java/com/sudomimus/token/IdToken.java
package com.sudomimus.token;

import com.fasterxml.jackson.databind.ObjectMapper;

import java.nio.charset.StandardCharsets;
import java.security.KeyFactory;
import java.security.PublicKey;
import java.security.Signature;
import java.security.spec.X509EncodedKeySpec;
import java.time.Instant;
import java.util.Base64;

/**
 * A parsed Sudomimus OIDC {@code id_token}. Unlike Sudomimus access/refresh
 * tokens (whose envelope claims live in the JWT header), an id_token is a
 * standard OIDC JWT: every claim lives in the body and the token is signed by
 * the <b>platform</b> key (resolve it from the OIDC JWKS by the header
 * {@code kid}), not by an application's signing key. {@link #verify} therefore
 * checks {@code exp} from the body.
 */
public final class IdToken {

    private static final ObjectMapper MAPPER = new ObjectMapper();
    private static final Base64.Decoder B64URL = Base64.getUrlDecoder();
    private static final String PEM_HEADER = "-----BEGIN PUBLIC KEY-----";
    private static final String PEM_FOOTER = "-----END PUBLIC KEY-----";

    private final String raw;
    private final byte[] signingInput;
    private final byte[] signature;
    private final IdTokenHeader header;
    private final IdTokenBody body;

    IdToken(String raw, byte[] signingInput, byte[] signature, IdTokenHeader header, IdTokenBody body) {
        this.raw = raw;
        this.signingInput = signingInput;
        this.signature = signature;
        this.header = header;
        this.body = body;
    }

    public String getRaw() { return raw; }
    public IdTokenHeader getHeader() { return header; }
    public IdTokenBody getBody() { return body; }

    /** Returns true when the RSA-SHA256 signature matches the given PEM-encoded public key. */
    public boolean verifySignature(String publicKeyPem) {
        try {
            PublicKey pub = parseRsaPublicKey(publicKeyPem);
            Signature sig = Signature.getInstance("SHA256withRSA");
            sig.initVerify(pub);
            sig.update(signingInput);
            return sig.verify(signature);
        } catch (Exception e) {
            return false;
        }
    }

    /** Parse an id_token into its header and body without verifying it. */
    public static IdToken parse(String jwt) {
        if (jwt == null || jwt.isEmpty()) {
            throw new TokenException(TokenErrorCode.INVALID_JWT, "Token is empty.");
        }
        String[] parts = jwt.split("\\.", -1);
        if (parts.length != 3) {
            throw new TokenException(TokenErrorCode.INVALID_JWT,
                    "Token must have exactly three dot-separated segments; got " + parts.length + ".");
        }

        IdTokenHeader header = decodeJson(parts[0], IdTokenHeader.class, "header");
        IdTokenBody body = decodeJson(parts[1], IdTokenBody.class, "body");

        byte[] signature;
        try {
            signature = B64URL.decode(parts[2]);
        } catch (IllegalArgumentException e) {
            throw new TokenException(TokenErrorCode.INVALID_JWT,
                    "Failed to decode id_token signature segment: " + e.getMessage(), e);
        }

        byte[] signingInput = (parts[0] + "." + parts[1]).getBytes(StandardCharsets.UTF_8);
        return new IdToken(jwt, signingInput, signature, header, body);
    }

    /**
     * Verify an OIDC id_token against a platform public key (resolved from the
     * OIDC JWKS). Checks the body {@code exp}, the RS256 signature, and any of
     * the supplied audience/issuer/nonce expectations.
     */
    public static IdToken verify(String jwt, String platformPublicKeyPem, IdTokenExpectations expectations) {
        IdToken parsed = parse(jwt);
        IdTokenExpectations expect = expectations != null ? expectations : new IdTokenExpectations();
        Instant now = expect.now != null ? expect.now : Instant.now();

        if (parsed.body.expiresAt == null
                || !now.isBefore(Instant.ofEpochSecond(parsed.body.expiresAt))) {
            throw new TokenException(TokenErrorCode.EXPIRED, "id_token has expired.");
        }

        if (!parsed.verifySignature(platformPublicKeyPem)) {
            throw new TokenException(TokenErrorCode.INVALID_SIGNATURE,
                    "id_token signature does not match the platform public key.");
        }

        if (expect.audience != null && !expect.audience.equals(parsed.body.audience)) {
            throw new TokenException(TokenErrorCode.WRONG_AUDIENCE,
                    "id_token aud does not match the expected client id.");
        }
        if (expect.issuer != null && !expect.issuer.equals(parsed.body.issuer)) {
            throw new TokenException(TokenErrorCode.WRONG_ISSUER,
                    "id_token iss does not match the expected issuer.");
        }
        if (expect.nonce != null && !expect.nonce.equals(parsed.body.nonce)) {
            throw new TokenException(TokenErrorCode.WRONG_NONCE,
                    "id_token nonce does not match the value sent at /authorize.");
        }

        return parsed;
    }

    private static <T> T decodeJson(String segment, Class<T> type, String label) {
        byte[] bytes;
        try {
            bytes = B64URL.decode(segment);
        } catch (IllegalArgumentException e) {
            throw new TokenException(TokenErrorCode.INVALID_JWT,
                    "Failed to decode id_token " + label + " segment: " + e.getMessage(), e);
        }
        try {
            T value = MAPPER.readValue(bytes, type);
            if (value == null) {
                throw new TokenException(TokenErrorCode.INVALID_JWT,
                        "id_token " + label + " deserialized to null.");
            }
            return value;
        } catch (TokenException e) {
            throw e;
        } catch (Exception e) {
            throw new TokenException(TokenErrorCode.INVALID_JWT,
                    "Failed to deserialize id_token " + label + ": " + e.getMessage(), e);
        }
    }

    private static PublicKey parseRsaPublicKey(String pem) throws Exception {
        String body = pem
                .replace(PEM_HEADER, "")
                .replace(PEM_FOOTER, "")
                .replaceAll("\\s+", "");
        byte[] der = Base64.getDecoder().decode(body);
        X509EncodedKeySpec spec = new X509EncodedKeySpec(der);
        return KeyFactory.getInstance("RSA").generatePublic(spec);
    }
}
